### PR TITLE
Extend team 1 agent

### PR DESCRIPTION
**Pull request by mpardalos**
## Summary

Extend team 1 agent:
* Custom president
* Opinion-based voting
* More config options!!!!11!!1111one

## Test Plan

All tests pass, some tests added for iigo code

### CODE REVIEW

File: internal/clients/team1/client.go [hunk 1]
```diff
@@ -3,16 +3,19 @@ package team1
...
+	"github.com/SOMAS2020/SOMAS2020/internal/common/disasters"
 	"github.com/SOMAS2020/SOMAS2020/internal/common/gamestate"
 	"github.com/SOMAS2020/SOMAS2020/internal/common/shared"
-	// "github.com/SOMAS2020/SOMAS2020/internal/clients/team1/foraging"
 )
 
 const id = shared.Team1
 
 type EmotionalState int
+type Opinion int
```
Comment (lhl2617) [inline internal/clients/team1/client.go:18]: dam, what the range of `Opinion`

File: internal/clients/team1/iifo.go [hunk 2]
```diff
@@ -56,3 +71,138 @@ func (c *client) ReceiveForageInfo(forageInfos []shared.ForageShareInfo) {
...
+	for _, disaster := range disasterHistory {
+		totalDisaster.X += math.Pow(disaster.X-meanDisaster.X, 2)
+		totalDisaster.Y += math.Pow(disaster.Y-meanDisaster.Y, 2)
+		totalDisaster.Magnitude += math.Pow(disaster.Magnitude-meanDisaster.Magnitude, 2)
+	}
+
+	disasterHistorySize := float64(len(disasterHistory))
+	sqrtDisasterHistory := math.Sqrt(disasterHistorySize)
+	xSD := math.Sqrt(totalDisaster.X / disasterHistorySize)
+	ySD := math.Sqrt(totalDisaster.Y / disasterHistorySize)
```
Comment (lhl2617) [inline internal/clients/team1/iifo.go:96]: div0

Comment (lhl2617) [reply internal/clients/team1/iifo.go:96]: below too

File: internal/clients/team1/client.go (hunk 1 above)
Comment (mpardalos) [reply internal/clients/team1/client.go:18]: -maxInt to +maxInt?

File: internal/clients/team1/iigo.go [hunk 3]
```diff
@@ -1,53 +1,179 @@
...
-		return 20
+	case Desperate, Anxious:
+		amount := shared.Resources(c.config.resourceRequestScale) *
+			c.gameConfig().CostOfLiving
+		fmt.Printf("Amount %v", amount)
+		c.Logf("Common pool request: %v", amount)
+		return amount
 	default:
-		return 0
+		return c.gameState().CommonPool
```
Comment (lhl2617) [inline internal/clients/team1/iigo.go:145]: oh ho ho ho ho ho ho

File: internal/clients/team1/foraging.go [hunk 4]
```diff
@@ -83,28 +83,44 @@ func (c client) regressionForage() (shared.ForageDecision, error) {
...
-		expectedReward := shared.Resources(expectedRewardF)
-
-		if expectedReward > bestReward {
-			bestReward = expectedReward
+		var expectedReward shared.Resources
+		var contribution shared.Resources
+		// Regression throws an error when the size of array is less than 2
+		if len(c.forageHistory[forageType]) < 2 {
+			if forageType == shared.DeerForageType {
+				c.Logf("[Forage decision] Ha! jokes. Flipping instead")
```
Comment (lhl2617) [inline internal/clients/team1/foraging.go:91]: LMAO

File: internal/clients/team1/iito.go [hunk 5]
```diff
@@ -141,17 +143,13 @@ func (c *client) SentGift(sent shared.Resources, to shared.ClientID) {
...
-				c.opinionTeams[i].opinion += c.config.maxOpinion / 5
-			} else if received > 0 {
-				c.opinionTeams[i].opinion++
-			} else if received <= 0 {
-				c.opinionTeams[i].opinion--
-			}
-		}
+	if received > shared.Resources(c.receivedOffer[from]) {
+		// We love them cause they gave more than they promised.
+		c.teamOpinions[from] += c.config.maxOpinion / 5
```
Comment (lhl2617) [inline internal/clients/team1/iito.go:148]: strange 5 but ok?

File: internal/clients/team1/client.go (hunk 1 above)
Comment (lhl2617) [reply internal/clients/team1/client.go:18]: ah ok kewl

File: internal/clients/team1/client.go [hunk 6]
```diff
@@ -113,10 +158,24 @@ func (c *client) StartOfTurn() {
 	c.Logf("Emotional state: %v", c.emotionalState())
 	c.Logf("Resources: %v", c.gameState().ClientInfo.Resources)
 
+	// This should only happen at the start of the game.
+	if c.gameState().Turn == 1 {
+		c.disasterInfo.meanDisaster = disasters.DisasterReport{}
+		c.forageType = shared.DeerForageType
+		if c.gameConfig().DisasterConfig.DisasterPeriod.Valid == true {
```
Comment (lhl2617) [inline internal/clients/team1/client.go]: linter

File: internal/clients/team1/iifo.go (hunk 2 above)
Comment (E-Mac27) [reply internal/clients/team1/iifo.go:96]: Never gonna be a disasterHistorySize == 0 but i'll add an if statement.

**Review by lhl2617 — APPROVED** [review]
yes let's go

**Pull request merged**